lib/spike.rb

```diff
@@ -0,0 +1,361 @@
+      @definition = JSON.parse(json, object_class: OpenStruct)
+      normalize!
+    end
+
+    private
+
+    def normalize!
+      @definition["original_path"] = definition.path
+      @definition["path"]          = normalized_path
+      @definition["params"]        = normalized_params
+
+      @definition["documentation_url"] = @definition["documentationUrl"]
+      @definition["verb"]              = @definition["method"]
+    end
+
+    def normalized_path
+      path = definition.path.to_s
+      if definition.path.to_s.include?(":id")
+        idx = original_path_segments.index(":id")
+        resource = original_path_segments[idx-1].singularize
+        path = path.gsub(":id", ":#{resource}_id")
+      end
+      path
+    end
+
+    def normalized_params
+      params = (@definition.params || []).reject do |param|
+        param.name == "owner"
+      end
+
+      params.each do |param|
+        if param.name == "id"
+          idx = original_path_segments.index(":id")
+          resource = original_path_segments[idx-1].singularize
+          param["name"] = "#{resource}_id"
+        end
+        if param.name.end_with?("_id") && param.type == "string"
+          param["type"] = "integer"
+        end
+        if param.name == "repo"
+          param["description"] = "A GitHub repository"
+        end
+        if param.description == "" && param.name.end_with?("_id")
+          param["description"] = "The ID of the #{param.name.gsub("_id", "").gsub("_", " ")}"
+        end
+        param.description = param.description.to_s.gsub(" Please see more in the alert below.", "")
+        if param.type.to_s.end_with?("[]")
+          param["type"] = "Array<#{param.type.gsub("[]", "")}>"
+        end
+      end
+
+      resource = Spike::Resource.new(definition.path)
+      if resource.subresource?
+        id_param = params.find {|param| param.name =~ /.*id/}&.name
+        if id_param
+          idx = path_segments.index(":#{id_param}")
+          resource = path_segments[idx-1].singularize
+          params = params.reject {|param|
+            ["repo", id_param].include?(param.name)
+          }
+          url_param = {
+            name: "#{resource}_url",
+            required: true,
+            type: "string",
+            description: "A URL for a #{resource} resource.",
+          }
+          params = params.unshift OpenStruct.new(url_param)
+        end
+      end
+
+      params
+    end
+
+    def path_segments
+      definition.path.to_s.split("/").reject(&:empty?)
+    end
+
+    def original_path_segments
+      definition.original_path.to_s.split("/")
+    end
+  end
+
+  class Endpoint
+    class PositionalParameterizer
+      def parameterize(args)
+        "#{args.join(", ")}, options = {}"
+      end
+    end
+
+    class KwargsParameterizer
+      def parameterize(args)
+        "#{args.map {|arg| arg + ":"}.join(", ")}, **options"
+      end
+    end
+
+    extend Forwardable
+
+    VERB_PRIORITY = %w(GET POST)
```

Comment:
I want to sort the client methods consistently throughout all the files:

* get a single thing
* list the things
* create a thing

and eventually we'll get the "update and delete" when we spike out another API that has them.